File: inventory/sources/pip.py
#!/usr/bin/env python3

import json
import logging
import sys

from ..inventory import InventoryItem
from .runner import run

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)


class PipSource:
    name = "Pip"
    items = []

    def collect(self) -> list[InventoryItem]:
        out = run(f"{sys.executable} -m pip list --format=json 2>/dev/null")
        if not out:
            return self.items

        try:
            pkgs = json.loads(out)
            for pkg in pkgs:
                self.items.append(
                    InventoryItem(
                        vendor="",  # Inferred or left blank
                        product=pkg["name"].lower().replace("-", "_"),
                        version=pkg["version"],
                        part="a",
                        source="pip",
                    )
                )
        except json.JSONDecodeError as e:
            logger.error(f"Failed to decode `pip list` results: {e}")
            return []

        return self.items

File: crates/lizt-cpe/src/sources/pip.rs
use crate::cpe::{CpeEntry, CpePart, CpeSource, SystemCpe};
use crate::inventory::Source;
use crate::runner::run;
use serde::Deserialize;
use tracing::error;

#[derive(Deserialize)]
struct PipPackage {
    name: String,
    version: String,
}

pub struct PipSource;

impl Source for PipSource {
    fn name(&self) -> &str {
        "pip"
    }

    fn collect(&self) -> Vec<SystemCpe> {
        let Some(out) = run("python3 -m pip list --format=json 2>/dev/null") else {
            return vec![];
        };

        match serde_json::from_str::<Vec<PipPackage>>(&out) {
            Ok(pkgs) => pkgs
                .into_iter()
                .map(|pkg| SystemCpe {
                    cpe: CpeEntry {
                        part: CpePart::Application,
                        vendor: String::new(),
                        product: pkg.name.to_lowercase().replace("-", "_"),
                        version: Some(pkg.version),
                        raw: String::new(), // Generated later
                    },
                    source: CpeSource::PackageManager(self.name().to_string()),
                })
                .collect(),
            Err(e) => {
                error!("Failed to decode `pip list` results: {}", e);
                vec![]
            }
        }
    }
}
